table.go

```diff
@@ -351,7 +362,140 @@ func (t *table) backfillTo() time.Time {
+		errCh:           make(chan error, 1),
+	}
+	t.db.requestedIterations <- it
+	return <-it.errCh
+}
+
+// coalesceIterations coalesces multiple parallel iterations for a single table,
+// which avoids having to do multiple table scans of the same table. It is
+// single-threaded, meaning that each ZenoDB instance performs only one scan at
+// a time. Since ZenoDB instances are meant to be sized small, this shouldn't be
+// a performance issue.
+func (db *DB) coalesceIterations() {
+	for it := range db.requestedIterations {
+		db.coalesceIteration(it)
+	}
+}
+
+func (db *DB) coalesceIteration(it *iteration) {
+	iterations := append([]*iteration(nil), it)
+	iterationsForOtherTables := make([]*iteration, 0)
+
+coalesceLoop:
+	for {
+		select {
+		case it2 := <-db.requestedIterations:
+			if it2.t == it.t {
+				iterations = append(iterations, it2)
+			} else {
+				iterationsForOtherTables = append(iterationsForOtherTables, it2)
+			}
+		case <-time.After(db.opts.IterationCoalesceInterval):
+			// stop waiting to coalesce
+			break coalesceLoop
+		}
+	}
+
+	// re-enqueue iterations for other tables since we won't be handling them
+	// here
+	for _, otherIt := range iterationsForOtherTables {
+		db.requestedIterations <- otherIt
+	}
+
+	var maxDeadline time.Time
+	includeMemStore := false
+	allOutFields := make(core.Fields, 0)
+	hasOutField := func(field core.Field) bool {
+		for _, existingField := range allOutFields {
+			if existingField.String() == field.String() {
+				return true
+			}
+		}
+		return false
+	}
+
+	for _, it := range iterations {
+		includeMemStore = includeMemStore || it.includeMemStore
+		deadline, hasDeadline := it.ctx.Deadline()
+		if hasDeadline && deadline.After(maxDeadline) {
+			maxDeadline = deadline
+		}
+		// default outFields to table fields
+		if it.outFields == nil {
+			it.outFields = it.t.fields
+		}
+		for _, field := range it.outFields {
+			if !hasOutField(field) {
+				allOutFields = append(allOutFields, field)
+			}
+		}
+	}
+
+	for _, it := range iterations {
+		it.fieldMappings = make(map[int]int, len(allOutFields))
+		for i, field := range allOutFields {
+			it.fieldMappings[i] = it.indexOfOutField(field)
+		}
+	}
+
+	log.Debugf("Coalescing %d iterations on %v", len(iterations), it.t.Name)
+
+	remainingIterations := make(map[int]*iteration, len(iterations))
+	for i, it := range iterations {
+		remainingIterations[i] = it
+	}
+
+	combinedOnValue := func(dims bytemap.ByteMap, vals []encoding.Sequence) (bool, error) {
+		more := false
+		for i, it := range remainingIterations {
+			itVals := make([]encoding.Sequence, len(it.outFields))
+			for i, val := range vals {
+				itI := it.fieldMappings[i]
+				if itI >= 0 {
+					itVals[itI] = val
+				}
+			}
+			itMore, err := it.onValue(dims, itVals)
+			if err != nil {
+				return false, err
+			}
+			if !itMore {
+				// This iteration doesn't want any more data, stop feeding it
+				delete(remainingIterations, i)
+			}
+			more = true
```

Comment:
else? or continue in the prior block?

Reply:
Nah, we're inside of a callback method that gets called once per actual row read from disk. This inner for loop is just distributing the results to the coalesced callbacks.

Reply:
I see that, but doesn't this more value control whether additional rows are visited in the iteration? It looks like this gets set to true even if everyone says false.  (I guess it just visits one more time and sees nobody so nbd?)